syntax = "proto3";

enum ExprKind {
    Lit = 0;
}

// Simplified representation of arbitrary expression
message Expr {
    ExprKind Kind = 1;
    string Value = 2;
    repeated Expr Sub = 3;
}

enum TypeKind {
    RegularType = 0;
    // Pointer to standalone function -- not pointer to member.
    // Return type is stored as a first argument, parameter types
    // are provided as trailing arguments
    FunctionPtr = 1;
    // Pointer to class method. Return type is stored as a first
    // parameter, class type is a second, argument are trailing. 
    MethodPtr = 2;
    Array = 3;
    TypeExpr = 4;
}

enum ReferenceKind {
    NotRef = 0;
    LValue = 1;
    RValue = 2;
}

enum TypeTag {
    TypeTagNone = 0;
    TypeTagStruct = 1;
    TypeTagEnum = 2;
    TypeTagUnion = 3;
}

message CVQual {
    bool IsPointer = 1;
    bool IsConst = 2;
    bool IsVolatile = 3;
}

// Fully qualified C++ type that is designed to uniquely identify the target
// within a single translation unit. More complex cases are not considered
// (duplicate type names in different files, across different projects)
message QualType {
    // Unmangled type name
    string Name = 1;
    // List of wrapping spaces that would allow to uniquely jump to the
    // type. If something was implicit in the code it will still be listed
    // fully here.
    repeated QualType Spaces = 2;
    // List of template type parameters for the type instance. Qualified types
    // that are used as parent types can also have parameters.
    repeated QualType Parameters = 3;
    repeated CVQual Qualifiers = 4;
    // `T&` vs `T&` (TODO r-value ref)
    ReferenceKind RefKind = 5;
    TypeKind Kind = 6;
    Expr TypeValue = 7;
    // `std::vector` -- std will also be handled as qualified type to
    // avoid making more complicated spaces structure. TODO revise if
    // `repeated string Spaces` + `repated QualType Parents` will be
    // a better alternative.
    bool IsNamespace = 8;
    string DbgOrigin = 9;
    TypeTag Tag = 10;
    bool IsBuiltin = 11;
}

// Enumeration definition
message Enum {
    message Field  {
        string Name = 1;
        int64 Value = 2;
    }

    repeated Field Fields = 1;
    QualType Name = 2;
    QualType Base = 3;
    string Doc = 4;
    /// In case enumeration was defined in C context, under both typedef and own name
    /// both are stored in the result. Generated data aims to have the most easily accessible
    /// 'name' for the enumeration declaration in the `Name` field, but if there is more data
    /// in the construction then it is pushed here. 
    QualType EnumDefName = 5;
    bool IsForwardDecl = 6;
}




message Arg {
    QualType Type = 1;
    string Name = 2;
    Expr Default = 3;
    string Doc = 4;
}


message TemplateTypename {
    bool IsVariadic = 1;
    bool IsPlaceholder = 2;
    string Name = 3;
    repeated TemplateTypename Nested = 4;
    string Concept = 5;
}

message TemplateSpec {
    repeated TemplateTypename Parameters = 1;
}

message Function {
    string Name = 1;
    QualType ResultTy = 2;
    repeated Arg Arguments = 3;
    repeated TemplateSpec Templates = 4;
    string DbgOrigin = 5;
    repeated QualType Spaces = 6;
}

message Typedef {
    /// Fully qualified name of the type definition -- trailing name and all parent scopes
    QualType Name = 1;
    /// Base type used in declaration
    QualType BaseType = 2;
    repeated TemplateSpec Templates = 3;
}

message Record {
    message Field {
        string Name = 1;
        string Doc = 2;
        QualType Type = 3;
        Record TypeDecl = 4;
        bool IsTypeDecl = 5;
    }

    enum MethodKind {
        Base = 0;
        CopyConstructor = 1;
        MoveConstructor = 2;
        CopyAssignmentOperator = 3;
        MoveAssignmentOperator = 4;
        Destructor = 5;
        DefaultConstructor = 6;
        ConvertingConstructor =  7;
    }

    message Method {
        QualType ReturnTy = 1;
        repeated Arg Args = 2;
        string Name = 3;
        bool IsConst = 4;
        bool IsStatic = 5;
        bool IsVirtual = 6 ;
        bool IsPureVirtual = 7;
        bool IsImplicit = 8;
        bool IsOperator = 9;
        MethodKind Kind = 10;
        string Doc = 11;
        string OperatorName = 12;
    }

    /// Name can be truly absent here, in case of `union {int a; char b};` that is
    /// used to splice inner members or for similar C-style approaches to the code 
    /// building. 
    QualType Name = 1;
    repeated QualType Bases = 2;
    repeated Field Fields = 3;
    repeated Method Methods = 4;
    repeated Record NestedRec = 5;
    repeated Enum NestedEnum = 6;
    repeated Typedef NestedTypedef = 7;
    string Doc = 8;
    repeated TemplateSpec Templates = 9;
    /// See 'EnumDefName' description in the enumeration message. 
    QualType RecordDefName = 10;
    bool IsForwardDecl = 11;
    /// Unions share most of the features with the regular records, so it is easier
    /// to introduce a side channel here. List of bases would be unnecessary, but 
    /// everything else can still be used in enum. 
    bool IsUnion = 12;
    /// Python protobuf generator I had to use to get type annotations does not support
    /// optional types (not important, I know), so I need to hack it around as a boolean
    /// field here. 
    bool HasName = 13;
    bool IsAbstract = 14;
}

message TU {
    repeated Record records = 1;
    repeated Enum enums = 2;
    repeated Function functions = 3;
    repeated Typedef typedefs = 4;
    string Doc = 5;
}
